server/src/main/java/org/elasticsearch/index/shard/ExplicitIndexSettingProvider.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.shard;
+
+import org.elasticsearch.common.settings.Settings;
+
+/**
+ * An {@link ExplicitIndexSettingProvider} is a provider for index level settings that can be set
+ * explicitly as a default value (so they show up as "set" for newly created indices)
+ */
+public interface ExplicitIndexSettingProvider {
+    /**
+     * Returns explicitly set default index {@link Settings} for the given index.
+     */
+    default Settings getExplicitIndexSettings(String indexName, Settings templateAndRequestSettings) {
```

Comment:
Can we do without the passed in settings? I had imagined the caller would do the merging/defaulting logic, so it's unclear why we would need the user passed in settings.

Reply:
We need the settings to be able to not set mutually exclusive settings, for example, if a user created an index with:

```
POST /myindex
{
  "settings": {
    "index.routing.allocation.require._name": "mynode"
  }
}
```

We need to *not* set `index.routing.allocation.include._tier` automatically, because otherwise we're constraining a new index to a tier when the user specifically wanted it constrained to single node (we check the all index level filtering settings). These settings are the only way we can make explicit default index settings reactive to other index level settings (such as adding an opt-out index level setting in the future)